src/producer.cpp

```diff
@@ -65,6 +65,42 @@ Producer::PayloadPolicy Producer::get_payload_policy() const {
 }
 
 void Producer::produce(const MessageBuilder& builder) {
+    do_produce(builder, MessageBuilder::HeaderListType(builder.header_list())); //copy headers
+}
+
+void Producer::produce(MessageBuilder&& builder) {
+    MessageBuilder temp(std::move(builder)); //owns header list after the move
+    do_produce(temp, MessageBuilder::HeaderListType(temp.header_list().release_handle())); //move headers
+}
+
+void Producer::produce(const Message& message) {
+    do_produce(message, HeaderList<Message::HeaderType>(message.get_header_list())); //copy headers
+}
+
+void Producer::produce(Message&& message) {
+    Message temp(std::move(message)); //rdakfka still owns the header list at this point
+    do_produce(temp, HeaderList<Message::HeaderType>(temp.detach_header_list<Message::HeaderType>())); //move headers
```

Comment:
I think the `HeaderList<Message::HeaderType>` is unnecessary, unless I'm missing something.

Reply:
You do need it, for `Message::get_header_list()` and `Message::detach_header_list()` are templetized on the type of header you want. `Message` itself has `Header<Buffer>` internally but there's no reason you can request a list as `Header<std::string>` out of it. For example, you can get the list as `Header<std::string>`, modify it and create a new builder from it.

Reply:
If you think that's too much overkill, I can remove the template and always force the user to get a `HeaderList<Header<Buffer>>` in which case if they want to modify it, they will need to create their own:
```
HeaderList<Header<std::string>> list(make_non_owning(message.detach_header_list().get_handle()));
```

Reply:
Yes you're right, I'll remove it.